packages/instantsearch.js/src/connectors/infinite-hits/connectInfiniteHits.ts

```diff
@@ -355,10 +356,31 @@ const connectInfiniteHits: InfiniteHitsConnector = function connectInfiniteHits(
             { results }
           );
 
+          /*
+            With dynamic widgets, facets are not included in the state before their relevant widgets are mounted. Until then, we need to bail out of writing this incomplete state representation in cache.
+          */
+          let hasDynamicWidgets = false;
+          walkIndex(instantSearchInstance.mainIndex, (indexWidget) => {
+            if (
+              !hasDynamicWidgets &&
+              indexWidget
+                .getWidgets()
+                .some(({ $$type }) => $$type === 'ais.dynamicWidgets')
+            ) {
+              hasDynamicWidgets = true;
+            }
+          });
+
+          const hasNoFacets =
+            !results?.disjunctiveFacets?.length &&
+            !results?.facets?.length &&
+            !results?.hierarchicalFacets?.length;
```

Comment:
Since you're in the `else` block of `if(!results)`, this should suffice :
```suggestion
            !results.disjunctiveFacets?.length &&
            !results.facets?.length &&
            !results.hierarchicalFacets?.length;
```